### PR TITLE
Add Permission System

### PR DESCRIPTION
**Pull request by InnayTool**
This adds permission checking for the notes route. The checking may be moved to an guard to make it cleaner.

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  signed-off my commits to accept the DCO.

### CODE REVIEW

File: src/permissions/permissions.service.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+@Injectable()
+export class PermissionsService {
+  constructor(private readonly logger: ConsoleLoggerService) {
+  }
+  mayRead(user : User, note : Note) : boolean {
+    if (this.isOwner(user, note)) return true;
+
+    if (this.hasPermissionUser(user, note, false)) return true;
+
+    if (this.hasPermissionUser(user, note, false)) return true;
```
Comment (ErikMichelson) [inline src/permissions/permissions.service.ts:21]: Despite this PR being still a draft, I want to point out this copy-and-paste error in case you didn't catch it.

```suggestion
    if (this.hasPermissionGroup(user, note, false)) return true;
```

Comment (InnayTool) [reply src/permissions/permissions.service.ts]: fixed

File: src/api/public/notes/notes.controller.ts [hunk 2]
```diff
@@ -54,20 +57,31 @@ export class NotesController {
     @MarkdownBody() text: string,
   ): Promise<NoteDto> {
     // ToDo: provide user for createNoteDto
-    this.logger.debug('Got raw markdown:\n' + text, 'createNote');
+    if (!this.permissionsService.mayCreate(req.user)) {
+      throw new UnauthorizedException('Creation denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:61]:
```suggestion
      throw new UnauthorizedException('Note creation denied!');
```

File: src/api/public/notes/notes.controller.ts [hunk 3]
```diff
@@ -54,20 +57,31 @@ export class NotesController {
...
+    }
+    this.logger.debug('Got raw markdown:\n' + text);
     return this.noteService.toNoteDto(
-      await this.noteService.createNote(text, undefined, req.user),
+      await this.noteService.createNote(text, '', req.user),
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:65]: Why did you change this? With this each Note will get the alias '' added.

File: src/api/public/notes/notes.controller.ts [hunk 4]
```diff
@@ -54,20 +57,31 @@ export class NotesController {
...
-  async getNote(@Request() req, @Param('noteIdOrAlias') noteIdOrAlias: string) {
-    // ToDo: check if user is allowed to view this note
+  async getNote(
+    @Request() req,
+    @Param('noteIdOrAlias') noteIdOrAlias: string,
+  ): Promise<NoteDto> {
     try {
       const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+      if (!this.permissionsService.mayRead(req.user, note)) {
+        throw new UnauthorizedException('Read denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:78]: see above

File: src/api/public/notes/notes.controller.ts [hunk 5]
```diff
@@ -83,7 +97,9 @@ export class NotesController {
     @Param('noteAlias') noteAlias: string,
     @MarkdownBody() text: string,
   ): Promise<NoteDto> {
-    // ToDo: check if user is allowed to view this note
+    if (!this.permissionsService.mayCreate(req.user)) {
+      throw new UnauthorizedException('Create denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:101]: see above

File: src/api/public/notes/notes.controller.ts [hunk 6]
```diff
@@ -96,7 +112,10 @@ export class NotesController {
     @Request() req,
     @Param('noteIdOrAlias') noteIdOrAlias: string,
   ): Promise<void> {
-    // ToDo: check if user is allowed to delete this note
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.isOwner(req.user, note)) {
+      throw new UnauthorizedException('Delete denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:117]: see above

File: src/api/public/notes/notes.controller.ts [hunk 7]
```diff
@@ -117,7 +136,10 @@ export class NotesController {
     @Param('noteIdOrAlias') noteIdOrAlias: string,
     @MarkdownBody() text: string,
   ): Promise<NoteDto> {
-    // ToDo: check if user is allowed to change this note
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.mayWrite(req.user, note)) {
+      throw new UnauthorizedException('Update denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:141]: see above

File: src/api/public/notes/notes.controller.ts [hunk 8]
```diff
@@ -138,7 +160,10 @@ export class NotesController {
     @Request() req,
     @Param('noteIdOrAlias') noteIdOrAlias: string,
   ): Promise<string> {
-    // ToDo: check if user is allowed to view this notes content
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.mayRead(req.user, note)) {
+      throw new UnauthorizedException('Read denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:165]: see above

File: src/api/public/notes/notes.controller.ts [hunk 9]
```diff
@@ -175,7 +203,10 @@ export class NotesController {
     @Param('noteIdOrAlias') noteIdOrAlias: string,
     @Body() updateDto: NotePermissionsUpdateDto,
   ): Promise<NotePermissionsDto> {
-    // ToDo: check if user is allowed to view this notes permissions
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.isOwner(req.user, note)) {
+      throw new UnauthorizedException('Update denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:208]: see above

File: src/api/public/notes/notes.controller.ts [hunk 10]
```diff
@@ -194,7 +225,10 @@ export class NotesController {
     @Request() req,
     @Param('noteIdOrAlias') noteIdOrAlias: string,
   ): Promise<RevisionMetadataDto[]> {
-    // ToDo: check if user is allowed to view this notes revisions
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.mayRead(req.user, note)) {
+      throw new UnauthorizedException('Raec denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:230]: see above
also typo

File: src/api/public/notes/notes.controller.ts [hunk 11]
```diff
@@ -219,7 +253,10 @@ export class NotesController {
     @Param('noteIdOrAlias') noteIdOrAlias: string,
     @Param('revisionId') revisionId: number,
   ): Promise<RevisionDto> {
-    // ToDo: check if user is allowed to view this notes revision
+    const note = await this.noteService.getNoteByIdOrAlias(noteIdOrAlias);
+    if (!this.permissionsService.mayRead(req.user, note)) {
+      throw new UnauthorizedException('Read denied!');
```
Comment (DerMolly) [inline src/api/public/notes/notes.controller.ts:258]: see above

File: src/notes/note.entity.ts [hunk 12]
```diff
@@ -70,9 +70,6 @@ export class Note {
   @JoinTable()
   tags: Tag[];
 
-  // eslint-disable-next-line @typescript-eslint/no-empty-function
-  private constructor() {}
-
```
Comment (DerMolly) [inline src/notes/note.entity.ts:75]: Should this really be used here?

File: src/permissions/permissions.service.ts [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+  mayCreate(user: User): boolean {
+    if (user) {
+      return true;
+    } else {
+      if (
+        this.guestPermission == GuestPermission.CREATE ||
+        this.guestPermission == GuestPermission.CREATE_ALIAS
+      ) {
+        // TODO change to guestPermission to config option
```
Comment (DerMolly) [inline src/permissions/permissions.service.ts:52]: Should this still be here?

File: src/permissions/permissions.service.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+          if (groupPermission.group.name == 'loggedIn') {
+            // TODO: Name of group for logged in users
+            return true;
+          }
+          if (
+            groupPermission.group.name == 'everybody' &&
+            (groupPermission.canEdit || !wantEdit) &&
+            guestsAllowed
+          ) {
+            // TODO: Name of group in which everybody even guests can edit
```
Comment (DerMolly) [inline src/permissions/permissions.service.ts:113]: Should this be still here?

**Review by DerMolly — CHANGES_REQUESTED** [review]

File: src/history/history.service.spec.ts [hunk 15]
```diff
@@ -99,7 +105,7 @@ describe('HistoryService', () => {
   describe('createOrUpdateHistoryEntry', () => {
     describe('works', () => {
       it('without an preexisting entry', async () => {
-        const user = new User();
+        const user = {} as User;
```
Comment (davidmehren) [inline src/history/history.service.spec.ts:108]: I'm not sure if this will haunt us when we try to enable strict mode.

File: src/monitoring/monitoring.service.ts [hunk 16]
```diff
@@ -16,6 +16,7 @@ let versionCache: null | {
   preRelease?: string;
   commit?: string;
 } = null;
+
```
Comment (davidmehren) [inline src/monitoring/monitoring.service.ts:19]: stray newline?

File: src/permissions/permissions.service.ts [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { Injectable } from '@nestjs/common';
+import { User } from '../users/user.entity';
+import { Note } from '../notes/note.entity';
+
+export enum GuestPermission {
```
Comment (davidmehren) [inline src/permissions/permissions.service.ts:11]: Shouldn't this be in a separate file?

File: src/permissions/permissions.service.ts [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+  DENY = 'deny',
+  READ = 'read',
+  WRITE = 'write',
+  CREATE = 'create',
+  CREATE_ALIAS = 'createAlias',
+}
+
+@Injectable()
+export class PermissionsService {
+  public guestPermission: GuestPermission; // TODO cange to configOption
```
Comment (davidmehren) [inline src/permissions/permissions.service.ts:21]: Typo in comment

**Review by davidmehren — CHANGES_REQUESTED** [review]
We should check if it is possible to move all the permission checks into a decorator to keep the code cleaner.

File: src/api/public/notes/notes.controller.ts (hunk 2 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 3 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 4 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 5 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 6 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 7 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 8 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 9 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 10 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/api/public/notes/notes.controller.ts (hunk 11 above)
Comment (InnayTool) [reply src/api/public/notes/notes.controller.ts]: done

File: src/notes/note.entity.ts (hunk 12 above)
Comment (InnayTool) [reply src/notes/note.entity.ts]: Added the constructor back again.

File: src/permissions/permissions.service.ts (hunk 13 above)
Comment (InnayTool) [reply src/permissions/permissions.service.ts:52]: Yes because config option is not defined by now.

File: src/permissions/permissions.service.ts (hunk 14 above)
Comment (InnayTool) [reply src/permissions/permissions.service.ts:113]: Yes it's not decided how to name this groups.

File: src/users/user.entity.ts [hunk 19]
```diff
@@ -52,6 +53,10 @@ export class User {
   @OneToMany((_) => Identity, (identity) => identity.user)
   identities: Identity[];
 
+  @ManyToMany((_) => Group, (group) => group.members)
+  groups : Group[];
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  private constructor() {}
```
[the remark this answers is not in the document]
Comment (InnayTool) [reply src/users/user.entity.ts]: done

File: src/permissions/permissions.service.ts (hunk 17 above)
Comment (InnayTool) [reply src/permissions/permissions.service.ts:11]: Yes but guest permissions have to be implemented in the config anyway.

File: src/monitoring/monitoring.service.ts (hunk 16 above)
Comment (InnayTool) [reply src/monitoring/monitoring.service.ts:19]: Yeah it's probably added by changing something and changing back again.

File: src/history/history.service.spec.ts (hunk 15 above)
Comment (InnayTool) [reply src/history/history.service.spec.ts:108]: Yes but what is the solution to mitigate that?

**Comment by InnayTool** [comment]
> We should check if it is possible to move all the permission checks into a decorator to keep the code cleaner.

I think it's possible but I think this should be added in a other PR.

File: src/history/history.service.spec.ts (hunk 15 above)
Comment (DerMolly) [reply src/history/history.service.spec.ts:108]: I think I might have found a solution. I added it to https://github.com/hedgedoc/hedgedoc/issues/835

File: src/permissions/permissions.service.ts (hunk 17 above)
Comment (davidmehren) [reply src/permissions/permissions.service.ts:11]: If you intend to move this to the config module at a later time, please add a todo.

File: src/history/history.service.spec.ts (hunk 15 above)
Comment (InnayTool) [reply src/history/history.service.spec.ts:108]: I think we should change this in another PR.

Comment (davidmehren) [reply src/history/history.service.spec.ts:108]: After the discussion in #835, I think it's best to leave this as is for now and think about alternatives when we enable strict mode.

**Review by davidmehren — DISMISSED** [review]

**Comment by davidmehren** [comment]
I'm generally happy with the code, but the tests don't implement permissions yet.
I'd like to avoid merging code with broken tests and heard that @molly wanted to rebuild the e2e tests.
The best way forward would probably be a separate PR against this branch which fixes the tests, after @InnayTool has cleaned up the commit history.

**Comment by DerMolly** [comment]
Well I wanted to fix the e2e test as part of my separate PR for completing the notes service. But tbh. I'm not to sure if those fixes are still in order as the mostly replaced service calls with api calls iirc. Let me check.

**Review by davidmehren — APPROVED** [review]

**Comment by InnayTool** [comment]
rebased

**Review by davidmehren — DISMISSED** [review]

**Review by DerMolly — APPROVED** [review]
